src/mame/roland/roland_d70.cpp

```diff
@@ -0,0 +1,562 @@
+}
+
+void roland_d70_state::d70_map(address_map &map) {
+  map(0x0100, 0x0100).w(FUNC(roland_d70_state::bank_w));
+  map(0x0400, 0x07ff).rw(FUNC(roland_d70_state::ksga_io_r), FUNC(roland_d70_state::ksga_io_w));
+  map(0x0800, 0x0800).rw(FUNC(roland_d70_state::lcd_ctrl_r), FUNC(roland_d70_state::lcd_ctrl_w));
+  map(0x0802, 0x0802).rw(FUNC(roland_d70_state::lcd_data_r), FUNC(roland_d70_state::lcd_data_w));
+  map(0x0900, 0x09ff).rw(FUNC(roland_d70_state::snd_io_r), FUNC(roland_d70_state::snd_io_w));
+  map(0x0a00, 0x0aff).rw(FUNC(roland_d70_state::dsp_io_r), FUNC(roland_d70_state::dsp_io_w));
+  map(0x0c00, 0x0cff).rw(FUNC(roland_d70_state::tvf_io_r), FUNC(roland_d70_state::tvf_io_w));
+  map(0x1000, 0x7FFF).rom().region("progrom", 0x1000);
+  map(0x8000, 0xbfff).bankrw("bank");
+  map(0xc000, 0xffff).bankrw("fixed");
+}
+
+void roland_d70_state::d70(machine_config &config) {
+  i8x9x_device &maincpu(N8097BH(config, cpu, 12_MHz_XTAL));
+  maincpu.set_addrmap(AS_PROGRAM, &roland_d70_state::d70_map);
+  // maincpu.serial_tx_cb().set(FUNC(roland_d70_state::midi_w));
+  maincpu.serial_tx_cb().set("mdout", FUNC(midi_port_device::write_txd));
+  maincpu.in_p0_cb().set(FUNC(roland_d70_state::port0_r));
+  maincpu.in_p1_cb().set(FUNC(roland_d70_state::port1_r));
+  maincpu.out_p1_cb().set(FUNC(roland_d70_state::port1_w));
+  maincpu.ach0_cb().set(FUNC(roland_d70_state::ach0_r));
+  maincpu.ach1_cb().set(FUNC(roland_d70_state::ach1_r));
+  maincpu.ach2_cb().set(FUNC(roland_d70_state::ach2_r));
+  maincpu.ach3_cb().set(FUNC(roland_d70_state::ach3_r));
+  maincpu.ach4_cb().set(FUNC(roland_d70_state::ach4_r));
+
+  RAM(config, ram).set_default_size("128K");
+
+  SPEAKER(config, "lspeaker").front_left();
+  SPEAKER(config, "rspeaker").front_right();
+
+  MB87419_MB87420(config, pcm, 32.768_MHz_XTAL);
+  pcm->int_callback().set_inputline(cpu, i8x9x_device::EXTINT_LINE);
+  pcm->add_route(0, "lspeaker", 1.0);
+  pcm->add_route(1, "rspeaker", 1.0);
+
+  RAM(config, dsp_ram).set_default_size("8K");
+
+  T6963C(config, lcd, 0);
+  lcd->set_addrmap(0, &roland_d70_state::lcd_map);
+
+  screen_device &screen(SCREEN(config, "screen", SCREEN_TYPE_LCD));
+  screen.set_refresh_hz(60);
+  screen.set_size(240, 64);
+  screen.set_visarea_full();
+  screen.set_screen_update("lcd", FUNC(t6963c_device::screen_update));
+  screen.set_palette("palette");
+
+  PALETTE(config, "palette", FUNC(roland_d70_state::lcd_palette), 2);
+
+  TIMER(config, midi_timer)
+      .configure_periodic(FUNC(roland_d70_state::midi_timer_cb),
+                          attotime::from_hz(1250));
+
+  TIMER(config, "samples_timer")
+      .configure_periodic(FUNC(roland_d70_state::samples_timer_cb),
+                          attotime::from_hz(32000 * 2));
+
+  midi_port_device &mdin(MIDI_PORT(config, "mdin", midiin_slot, "midiin"));
+  mdin.rxd_handler().set([this](int state) {
+    // logerror("midi bit %x pos %x\n", state, m_midi_pos);
+
+    if (m_midi_pos == 0 || m_midi_pos == 9) {
```

Comment:
This isn't ideal but it's not your fault since the CPU device doesn't have proper serial yet.